website/docs/language/modules/develop/refactoring.html.md

```diff
@@ -0,0 +1,461 @@
+}
+```
+
+The original module, now only a shim for backward-compatibility, calls the
+two new modules and indicates that the resources moved into them:
+
+```hcl
+module "x" {
+  source = "../modules/x"
+
+  # ...
+}
+
+module "y" {
+  source = "../modules/y"
+
+  # ...
+}
+
+moved {
+  from = aws_instance.a
+  to   = module.x.aws_instance.a
+}
+
+moved {
+  from = aws_instance.b
+  to   = module.x.aws_instance.b
+}
+
+moved {
+  from = aws_instance.c
+  to   = module.y.aws_instance.c
+}
+```
+
+When an existing user of the original module upgrades to the new "shim"
+version, Terraform will notice these three `moved` blocks and thus behave
+as if the objects associated with the three old resource addresses were
+originally created inside the two new modules.
+
+New users of this family of modules may use either the combined shim module
+_or_ the two new modules separately. You may wish to communicate to your
+existing users that the old module is now deprecated and so they should use
+the two separate modules for any new needs.
+
+The multi-module refactoring situation is unusual in that it violates the
+typical rule that a parent module sees its child module as a "closed box",
+unaware of exactly which resources are declared inside it. This compromise
+assumes that all three of these modules are maintained by the same people
+and distributed together in a single
+[module package](../sources.html#modules-in-package-sub-directories).
+
+To avoid strong [coupling](https://en.wikipedia.org/wiki/Coupling_(computer_programming))
+between separately-packaged modules, Terraform will only allow declarations of
+moves between modules in the same package. In other words, Terraform would
+not have allowed moving into `module.x` above if the `source` address of
+that call had not been a [local path](../sources.html#local-paths).
+
+Terraform resolves module references in `moved` blocks relative to the module
+instance they are defined in. For example, if the original module above were
+already a child module named `module.original`, the reference to
+`module.x.aws_instance.a` would resolve as
+`module.original.module.x.aws_instance.a`. A module may only make `moved`
+statements about its own objects and objects of its child modules.
+
+If you need to refer to resources within a module that was called using
+`count` or `for_each` meta-arguments, you must specify a specific instance
+key to use in order to match with the new location of the resource
+configuration:
+
+```hcl
+moved {
+  from = aws_instance.example
+  to   = module.new[2].aws_instance.example
+}
+```
+
+## Removing `moved` Blocks
+
+Over time, a long-lasting module may accumulate many `moved` blocks.
+
+If you are maintaining private modules within an organization and you know
+for certain that all users of your module have successfully run
+`terraform apply` with your new module version already then it can be safe
+to remove `moved` blocks in later versions of your module.
```

Comment:
Where possible, I try to rework sentences with a long "if" clause at the beginning. Flipping them around like this generally makes them easier to read :-) 

```suggestion
It can be safe to remove `moved` blocks in later versions of your module when you are maintaining private modules within an organization and all users of your module have successfully run
`terraform apply` with your new module version.
```

Reply:
This restructuring looks good to me. Thanks!

One way your new paragraph feels different to my original intent is that I was using "know for certain" (admittedly, clumsily) to try to suggest that this is something that the person making the decision needs to actively check, rather than just to _assume_.

Perhaps that's not really so important as I originally thought, but the consequence of getting this wrong is that a user of the module that _hadn't_ yet run `terraform apply` with a version including the `moved` block will see a plan to _delete_ the object that was moved, and applying that plan could be hard to recover from if the object is something like a database or some similar server which has state embedded in it.

I'm not super set on including this, but I wonder if you have some thoughts about this.

Reply:
When I was revising this I changed that part to say "...when you are maintaining private modules within an organization and _you know that_ all module users have...", but otherwise took this exactly as you wrote it.

I'm happy to still drop the extra words if you think they are unnecessary!

Reply:
yeah that's fair - I actually went back and forth about that. It's minor and I think a little extra prompting never hurt anyone :-) I agree with what you did!

Reply:
Thanks!